presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,31 +79,349 @@
+     * close() method, and in doing so closes their HTTP clients. If anybody is
+     * still trying to use a PrestoConnection created by the driver, they're
+     * SOL.
+     */
+    private static class ReferenceCountingLoadingCache<K, V>
+    {
+        private class Holder
+        {
+            private final V value;
+            private int refcount;
+            private int cleanerCount;
+            private ScheduledFuture currentCleanup;
+
+            private Holder(V value)
+            {
+                this.value = requireNonNull(value, "value is null");
+            }
+
+            private V get()
+            {
+                return value;
+            }
+
+            private int getRefcount()
+            {
+                return refcount;
+            }
+
+            private int getCleanerCount()
+            {
+                return cleanerCount;
+            }
+
+            private int reference()
+            {
+                setCleanup(null);
+                return ++refcount;
+            }
+
+            private int dereference()
+            {
+                return --refcount;
+            }
+
+            private void setCleanup(ScheduledFuture cleanup)
+            {
+                if (currentCleanup != null && currentCleanup.cancel(false)) {
+                    --cleanerCount;
+                }
+
+                checkState(cleanerCount >= 0, "Negative cleanerCount in setCleanup");
+
+                currentCleanup = cleanup;
+                if (cleanup != null) {
+                    ++cleanerCount;
+                }
+            }
+
+            private void scheduleCleanup(
+                    ScheduledExecutorService cleanupService,
+                    Runnable cleanupTask,
+                    long delay,
+                    TimeUnit unit)
+            {
+                checkState(refcount == 0, "non-zero refcount in scheduleCleanup");
+                ScheduledFuture cleanup = cleanupService.schedule(cleanupTask, delay, unit);
+                setCleanup(cleanup);
+            }
+
+            private void releaseForCleaning()
+            {
+                --cleanerCount;
+                checkState(cleanerCount >= 0, "Negative cleanerCount in releaseForCleaning");
+            }
+        }
+
+        private final ScheduledExecutorService valueCleanupService;
+        private final LoadingCache<K, Holder> backingCache;
+        private final AtomicBoolean closed = new AtomicBoolean(false);
+
+        // TODO: This is probably on the long side. Maybe change to 30 seconds?
+        private static final long retentionPeriod = 2;
+        private static final TimeUnit retentionUnit = TimeUnit.MINUTES;
+
+        private ReferenceCountingLoadingCache(CacheLoader<K, V> loader, Consumer<V> disposer)
+        {
+            this.valueCleanupService = new ScheduledThreadPoolExecutor(1, daemonThreadsNamed("cache-cleanup-%s"));
```

Comment:
No need for the number in the name since there's only one thread.
```java
this.valueCleanupService = newSingleThreadScheduledExecutor(daemonThreadsNamed("cache-cleanup"));